### PR TITLE
Add dir-diff support and various diff options

### PR DESCRIPTION
**Pull request by jhenkens**
Looking for feedback on the below changes - I don't think they are ready to merge in, but looking for early advice (or taking over ownership, if there is an eager maintainer).

The dir-diff support is in relation to https://github.com/gitextensions/gitextensions/issues/1647
The diff modes are improvements I wanted while working on the dir-diff support.

Fixes #

Changes proposed in this pull request:
- Add support for diff-tool, in the settings page
- Add two new compare modes 
  - Compare to Working Directory
  - Compare Selected Commits (use CTRL + click to select two commits, compare them in the order selected)

 
Screenshots before and after (if PR changes UI):
image
image

What did I do to test the code and ensure quality:
- Local testing with my setup

Has been tested on (remove any that don't apply):
- GIT 2.17
- Windows 10

### CODE REVIEW

**Comment by vbjay** [comment]
Fyi. To signify a pull request as not ready just add WIP to tye beginning
of the title. It stands for work in progress.

**Comment by drewnoakes** [comment]
Looks good at a glance, though I don't know much about diff tools so won't offer to be a primary reviewer.

Can you elaborate on why you feel this isn't ready for merging?

**Comment by jhenkens** [comment]
No unit testing, very limited testing of clients, known issue https://github.com/gitextensions/gitextensions/pull/5025/files#diff-05956b8e9b35344894a0ffe609cf32b9R3447 that I'm not certain is a local problem, or a tooling problem.

That issue manifests when clicking Working Directory, then compare to current branch.

File: GitUI/UserControls/RevisionGrid.cs [hunk 1]
```diff
@@ -3644,7 +3648,7 @@ private void CompareToBranchToolStripMenuItem_Click(object sender, EventArgs e)
                 if (form.ShowDialog(this) == DialogResult.OK)
                 {
                     var baseCommit = Module.RevParse(form.BranchName);
-                    UICommands.ShowFormDiff(IsFirstParentValid(), baseCommit, headCommit.ObjectId,
+                    UICommands.PerformDiff(IsFirstParentValid(), baseCommit, headCommit.ObjectId,
```
Comment (gerhardol) [inline GitUI/UserControls/RevisionGrid.cs]: Should this really be changed?

File: GitCommands/Git/GitModule.cs [hunk 2]
```diff
@@ -3444,6 +3444,15 @@ public IEnumerable<string> GetPreviousCommitMessages(int count, string revision
                 });
         }
 
+        // Known bug on git-for-windows: Comparing working directory to any branch, fails, due to -R
+        // I.e., git difftool --gui --no-prompt --dir-diff -R HEAD fails, but
```
Comment (gerhardol) [inline GitCommands/Git/GitModule.cs]: If so, information about the diff need to be transferred to _revisionDiffProvider.Get()
The -R option is used to display diffs correctly when comparing staged/unstaged artificial commits to other commits.

File: GitUI/UserControls/RevisionGrid.cs (hunk 1 above)
Comment (jhenkens) [reply GitUI/UserControls/RevisionGrid.cs]: I'm not sure the issue here. I presume we would want to be able to perform a directory diff from "Compare to branch" popup dialog, would we not?

Or is the comment in regards to the general renaming from ShowFormDiff to PerformDiff. That was a decision I made to clearly distiguish that this may not show a GitExtensions form. PerformDiff encompases the logic that either launches the dir-diff tool directly, or shows a GitExtensions form.

File: GitCommands/Git/GitModule.cs (hunk 2 above)
Comment (jhenkens) [reply GitCommands/Git/GitModule.cs]: First - I am but one person using this application. I did not know if this "known bug" is actually a bug for all, or a bug for me. I'd appreciate help testing this.

Second, I understand the need for the -R, but I was unsure as to what to do in this case, if the git-for-windows bug is truly broken for everyone. I could add logic to _revisionDiffProvider to know whether or not it is a dir-diff, and open a backwards diff. Or fall back to the GitExtensions display, file by file.

**Comment by RussKie** [comment]
Does this bug affect all versions of gfw or only specific ones? Has it been
fixed?
E.g. if it was fixed in the min required version (2.9.0?) then we don't
need to worry about it...

**Comment by jhenkens** [comment]
I guess I'll clarify. It's a bug that I have found, but not found any other mentions of. I don't think its known to gfw developers as of yet. I've reproduced it on two systems, one with GFW 2.17.0, another with 2.17.1.windows.2

**Comment by jhenkens** [comment]
Unless I hear otherwise, I think the best course of action is #2 - "disable comparing "From" the working directory when --dir-diff is enabled. One could still compare "to" the working directory, via the new "Compare to Working Directory" button, or via "Select as base" - "Compare to base"".

I will tackle this later this week and clean everything else up.

**Comment by jhenkens** [comment]
Any further thoughts or blockers?

File: GitUI/UserControls/RevisionGrid.cs [hunk 3]
```diff
@@ -991,11 +991,26 @@ private void RevisionsSelectionChanged(object sender, EventArgs e)
...
+            // I.e., git difftool --gui --no-prompt --dir-diff -R HEAD fails, but
+            // git difftool --gui --no-prompt --dir-diff HEAD succeeds
+            // Thus, we disable comparing "from" working directory.
+            var enabledDiffOnWorkingDirectory = !(AppSettings.UseDifftoolDirDiff && firstSelectedRevision?.Guid == GitRevision.UnstagedGuid);
+
+            compareSelectedCommitsMenuItem.Enabled = selectedRevisions.Count == 2 && firstSelectedRevision != null && secondSelectedRevision != null && enabledDiffOnWorkingDirectory;
+
+            compareToBranchToolStripMenuItem.Enabled = enabledDiffOnWorkingDirectory;
+            compareWithCurrentBranchToolStripMenuItem.Enabled = enabledDiffOnWorkingDirectory;
+            compareSelectedCommitsMenuItem.Enabled = enabledDiffOnWorkingDirectory;
```
Comment (gerhardol) [inline GitUI/UserControls/RevisionGrid.cs]: only use check above. Exception with builtin diff if only one is selected

File: GitUI/CommandsDialogs/SettingsDialog/Pages/DiffViewerSettingsPage.Designer.cs [hunk 4]
```diff
@@ -51,6 +52,7 @@ private void InitializeComponent()
             this.tableLayoutPanelForDiffViewer.ColumnStyles.Add(new System.Windows.Forms.ColumnStyle());
             this.tableLayoutPanelForDiffViewer.ColumnStyles.Add(new System.Windows.Forms.ColumnStyle());
             this.tableLayoutPanelForDiffViewer.Controls.Add(this.chkShowDiffForAllParents, 0, 6);
+            this.tableLayoutPanelForDiffViewer.Controls.Add(this.useDifftoolDirDiff, 0, 8);
```
Comment (gerhardol) [inline GitUI/CommandsDialogs/SettingsDialog/Pages/DiffViewerSettingsPage.Designer.cs]: This is not really related to the diff viewer, but there are not really better place for it. Advanced/Detailed need to be rearranged...

File: GitUI/UserControls/RevisionGrid.cs [hunk 5]
```diff
@@ -991,11 +991,26 @@ private void RevisionsSelectionChanged(object sender, EventArgs e)
...
+                compareToWorkingDirectoryMenuItem.Enabled = firstSelectedRevision.ObjectId != ObjectId.UnstagedId;
             }
 
+            // Bug in git-for-windows: Comparing working directory to any branch, fails, due to -R
+            // I.e., git difftool --gui --no-prompt --dir-diff -R HEAD fails, but
+            // git difftool --gui --no-prompt --dir-diff HEAD succeeds
+            // Thus, we disable comparing "from" working directory.
+            var enabledDiffOnWorkingDirectory = !(AppSettings.UseDifftoolDirDiff && firstSelectedRevision?.Guid == GitRevision.UnstagedGuid);
+
+            compareSelectedCommitsMenuItem.Enabled = selectedRevisions.Count == 2 && firstSelectedRevision != null && secondSelectedRevision != null && enabledDiffOnWorkingDirectory;
```
Comment (gerhardol) [inline GitUI/UserControls/RevisionGrid.cs]: Most behavior (at least in this menu) is allowing more than on revisions when selecting (the last two clicked are used). This changes this slightly

File: GitUI/CommandsDialogs/SettingsDialog/Pages/DiffViewerSettingsPage.Designer.cs (hunk 4 above)
Comment (jhenkens) [reply GitUI/CommandsDialogs/SettingsDialog/Pages/DiffViewerSettingsPage.Designer.cs]: Agreed, feels weird here, but didn't change.

File: GitUI/UserControls/RevisionGrid.cs [hunk 6]
```diff
@@ -3766,10 +3781,34 @@ private void compareToBaseToolStripMenuItem_Click(object sender, EventArgs e)
...
+        }
+
+        private void compareSelectedCommitsMenuItem_Click(object sender, EventArgs e)
+        {
+            var revisions = GetSelectedRevisions();
+            var headCommit = revisions.First();
+            var baseCommit = revisions.Skip(1)
+                .First();
+
+            UICommands.PerformDiff(IsFirstParentValid(), baseCommit.ObjectId, headCommit.ObjectId,
```
[the remark this answers is not in the document]
Comment (jhenkens) [reply GitUI/UserControls/RevisionGrid.cs]: See above.

File: GitUI/UserControls/RevisionGrid.cs [hunk 7]
```diff
@@ -3766,10 +3781,34 @@ private void compareToBaseToolStripMenuItem_Click(object sender, EventArgs e)
...
+        private void compareToWorkingDirectoryMenuItem_Click(object sender, EventArgs e)
+        {
+            var baseCommit = GetSelectedRevisions().First();
+            if (baseCommit.ObjectId == ObjectId.UnstagedId)
+            {
+                MessageBox.Show(this, "Cannot diff working directory to itself");
+                return;
+            }
+
+            UICommands.PerformDiff(IsFirstParentValid(), baseCommit.ObjectId, ObjectId.UnstagedId,
```
[the remark this answers is not in the document]
Comment (jhenkens) [reply GitUI/UserControls/RevisionGrid.cs]: This was an error about trying to find a merge-base for the compare to merge base checkbox. I made that conditional based on whether or not a merge base could be found, and disabled the checkbox if a merge base could not.

File: GitUI/UserControls/RevisionGrid.cs (hunk 3 above)
Comment (jhenkens) [reply GitUI/UserControls/RevisionGrid.cs]: Bug. Fixed.

File: GitUI/UserControls/RevisionGrid.cs [hunk 8]
```diff
@@ -991,11 +991,26 @@ private void RevisionsSelectionChanged(object sender, EventArgs e)
 
             var selectedRevisions = GetSelectedRevisions();
             var firstSelectedRevision = selectedRevisions.FirstOrDefault();
+            var secondSelectedRevision = selectedRevisions.Skip(1).FirstOrDefault();
             if (selectedRevisions.Count == 1 && firstSelectedRevision != null)
             {
                 _navigationHistory.Push(firstSelectedRevision.Guid);
+                compareToWorkingDirectoryMenuItem.Enabled = firstSelectedRevision.ObjectId != ObjectId.UnstagedId;
```
[the remark this answers is not in the document]
Comment (jhenkens) [reply GitUI/UserControls/RevisionGrid.cs]: Fixed.

File: GitUI/UserControls/RevisionGrid.cs (hunk 5 above)
Comment (jhenkens) [reply GitUI/UserControls/RevisionGrid.cs]: I believe I've changed this to work as you mention.

**Comment by jhenkens** [comment]
Missed the general feedback.

>Some comments about details, but I have a more general comment: I would expect dir-diff to be available together with FormDiff, not either or. Maybe the new menu options should be dedicated to dir-diff. (With the updates to artificial revisions, I hardly use FormDiff though.)

I think that they both provide essentially the same functionality in two very separate user interfaces. As someone who is quite used to a "dir-diff" workflow, I'd say having them comingled, or a second button click away (from within the "FormDiff") window adds a good bit of extra clicks needed to get what is desired, done.

**Comment by jhenkens** [comment]
@gerhardol I think I have addressed your concerns and responded to your overall discussion. I have just merged up with the latest refactors that have happened on default, and reapplied some changes that were initially in conflict. Please let me know what you think

**Comment by jhenkens** [comment]
Ah - I do now see a flaw with the "getting rid of the form-diff" window. It also eliminates the "compare to merge base" possibility, which is actually quite critical a lot of the time when diffing.

I'd love some suggestions on how to do this elegantly, from anyone involved. I feel the combination of:
`Right Click -> Dropdown -> Click` is already somewhat heavy (but I'll accept it as its what is already here).

But `Right` Click -> Dropdown -> Click -> Click Checkbox -> Click Open in directory diff tool` just seems horrendous.

**Comment by jhenkens** [comment]
Added customizable shortcuts to the dropdowns for compare, and also added mnemonics to the form diff buttons, so that one can easily navigate the compare workflows I've describe entirely via keyboard. 

I think the changeset is now in a good spot.

File: GitUI/CommandsDialogs/FormDiff.cs [hunk 9]
```diff
@@ -82,6 +86,13 @@ private void PopulateDiffFiles()
             DiffFiles.SetDiffs(ckCompareToMergeBase.Checked
                 ? new[] { _headRevision, _mergeBase }
                 : new[] { _headRevision, _baseRevision });
+
+            // Bug in git-for-windows: Comparing working directory to any branch, fails, due to -R
+            // I.e., git difftool --gui --no-prompt --dir-diff -R HEAD fails, but
+            // git difftool --gui --no-prompt --dir-diff HEAD succeeds
+            // Thus, we disable comparing "from" working directory.
```
Comment (gerhardol) [inline GitUI/CommandsDialogs/FormDiff.cs]: Open with difftool button is still enabled

File: GitUI/UserControls/RevisionGrid/RevisionGridControl.Designer.cs [hunk 10]
```diff
@@ -232,7 +235,6 @@ private void InitializeComponent()
             // 
             this.createNewBranchToolStripMenuItem.Image = global::GitUI.Properties.Images.BranchCreate;
             this.createNewBranchToolStripMenuItem.Name = "createNewBranchToolStripMenuItem";
-            this.createNewBranchToolStripMenuItem.ShortcutKeys = ((System.Windows.Forms.Keys)((System.Windows.Forms.Keys.Control | System.Windows.Forms.Keys.B)));
```
Comment (gerhardol) [inline GitUI/UserControls/RevisionGrid/RevisionGridControl.Designer.cs]: Note: Not in this PR, the shortcut keys removed in this file are not in Settings

Comment (jhenkens) [reply GitUI/UserControls/RevisionGrid/RevisionGridControl.Designer.cs]: This shortcut still works without issue from the menu-bar hotkey, which is reassignable via the settings page.

File: GitUI/CommandsDialogs/FormDiff.cs [hunk 11]
```diff
@@ -219,6 +230,16 @@ private void ckCompareToMergeBase_CheckedChanged(object sender, EventArgs e)
             PopulateDiffFiles();
         }
 
+        private void setCkCompareToMergeBaseText(string commit)
+        {
+            ckCompareToMergeBase.Text = $"Compare to merge &base ({commit})";
```
[the remark this answers is not in the document]
Comment (jhenkens) [reply GitUI/CommandsDialogs/FormDiff.cs]: I made the change, but it eliminates the mnemonic, sadly. Not sure why, but don't think it was worth my time.

**Comment by jhenkens** [comment]
I don't see a good pattern for ways to test this based on existing tests.

The closest existing tests I could find were for the GitCommands. Testing the git difftool --dir-diff command is difficult - it opens an external tool, only. Not sure its a great candidate. The rest are function UI tests, the closest equivalents I could find were model tests on a few helper models.

**Comment by gerhardol** [comment]
> I don't see a good pattern for ways to test this based on existing tests.

@RussKie 
Comments on tests here?
I do not believe that tests add value for these changes (other than increasing the reported test coverage).

**Comment by RussKie** [comment]
Please rebase on top of the current master and be ready to squash.

**Comment by RussKie** [comment]
image

I guess P4merge isn't capable of comparing directories... 😯

File: GitUI/CommandsDialogs/FormDiff.Designer.cs [hunk 12]
```diff
@@ -166,15 +170,39 @@ private void InitializeComponent()
...
+            this.ckCompareToMergeBase.TabIndex = 9;
+            this.ckCompareToMergeBase.Text = "Compare to merge &base";
+            this.ckCompareToMergeBase.UseMnemonic = true;
             this.ckCompareToMergeBase.UseVisualStyleBackColor = true;
             this.ckCompareToMergeBase.CheckedChanged += new System.EventHandler(this.ckCompareToMergeBase_CheckedChanged);
             // 
+            // btnCompareDirectoriesWithDiffTool
+            // 
+            this.btnCompareDirectoriesWithDiffTool.Anchor = AnchorStyles.Left;
+            this.btnCompareDirectoriesWithDiffTool.AutoSize = true;
```
Comment (RussKie) [inline GitUI/CommandsDialogs/FormDiff.Designer.cs]: Need to add `AutoSizeMode = GrowAndShrink`

**Comment by gerhardol** [comment]
> I guess P4merge isn't capable of comparing directories...

Beyond Compare is good at all compares.
For free applications, at least WinMerge (at least betas) support this.

**Comment by drewnoakes** [comment]
Meld is a decent FOSS option.

File: GitUI/CommandsDialogs/FormDiff.cs (hunk 9 above)
Comment (jhenkens) [reply GitUI/CommandsDialogs/FormDiff.cs]: Yes - the issue was `            openWithDifftoolToolStripMenuItem.Enabled = enableDifftoolDirDiff;` rather than `            btnCompareDirectoriesWithDiffTool.Enabled = enableDifftoolDirDiff;`. A bug.

File: GitUI/CommandsDialogs/FormDiff.cs [hunk 13]
```diff
@@ -219,6 +231,16 @@ private void ckCompareToMergeBase_CheckedChanged(object sender, EventArgs e)
             PopulateDiffFiles();
         }
 
+        private void setCkCompareToMergeBaseText(string commit)
```
[the remark this answers is not in the document]
Comment (jhenkens) [reply GitUI/CommandsDialogs/FormDiff.cs]: ✔️

**Comment by jhenkens** [comment]
@RussKie Perhaps I should add back in the options page, to allow a user to checkbox whether or not their diff tool supports directory diffing?

File: GitUI/CommandsDialogs/FormDiff.Designer.cs [hunk 14]
```diff
@@ -166,15 +170,39 @@ private void InitializeComponent()
...
+            this.btnCompareDirectoriesWithDiffTool.Name = "btnCompareDirectoriesWithDiffTool";
+            this.btnCompareDirectoriesWithDiffTool.Margin = new System.Windows.Forms.Padding(0);
+            this.btnCompareDirectoriesWithDiffTool.Size = new System.Drawing.Size(141, 17);
+            this.btnCompareDirectoriesWithDiffTool.TabIndex = 20;
```
[the remark this answers is not in the document]
Comment (jhenkens) [reply GitUI/CommandsDialogs/FormDiff.Designer.cs]: New to winforms - is there anything TabIndex does besides set the order over movement via cursor for tabs? I notice TabIndex set on a lot of components that are not selectable (panels, groups, etc), as well as the main diff panel, which once entered you cannot leave via tab. To me, the tab index on this entire page feels weird, but maybe its a PEBKAC.

File: GitUI/CommandsDialogs/FormDiff.Designer.cs (hunk 12 above)
Comment (jhenkens) [reply GitUI/CommandsDialogs/FormDiff.Designer.cs]: ✔️

**Comment by vbjay** [comment]
Click view tab order when in the form designer. You'll get the idea pretty
quickly.

**Comment by jhenkens** [comment]
Ah - the things I didn't know, as a Rider based developer. I'll give it a shot.

**Comment by gerhardol** [comment]
> Last question is posted a few comments - re the options page.

The diff-tool popup is good enough for me - the user should know that they have tried a directory diff and decipher the error message. There are enough options in GE already (even if I personally like to be able to change details). The diff/merge tool settings should be rewritten anyway. I vote for no option.

That said, I fully understand if someone else request the option.

**Comment by jhenkens** [comment]
Hate to be that-guy, but ping?

Any other comments?

**Comment by vbjay** [comment]
image

**Comment by jhenkens** [comment]
image

The only commoent "see review" sends me to _was addressed by me_. Everything else from that review was also addressed.

image

As well as updating the tab indexes via Visual Studio to make sense.

**Comment by RussKie** [comment]
Apologies, I haven't forgotten about it. I'm absolutely flat out busy with
family and work commitments at present.
I'll review the changes as soon as I can.

**Comment by RussKie** [comment]
Thank you
